### PR TITLE
Pure python implementation of the Trapezoidal method

### PR DESCRIPTION
**Pull request by ojha-aditya**
# Trapezoidal method implemented using direct implementation of the algorithm (#2)

## Linting
  - The code lints perfectly with a pylint score of 10

## Contents
  - calculus.py and test_calculus.py modified to include the pure algorithmic implementation of Trapezoidal rule.
  - calculus.py and test_calculus.py
  - function to implement Trapezoidal method trapezoid_python() in calculus.py and its corresponding unit test as per #2.

## Labels and Milestones
  - Label: Algorithm; Milestone: Algorithm - Trapezoidal Method of Numerical Integration

## Reviewer
  - @SchrodingersStruggle, @HeshamElsaman, @dhamalakamal, @zbpetersbuf or anyone else willing to review.

### CODE REVIEW

**Comment by SchrodingersStruggle** [comment]
Sorry that I didn't get the chance to review yesterday, but I can review this after I get out of a meeting, which would be at 6:00pm today.

**Comment by ojha-aditya** [comment]
Sure thing! I will sync this with the main branch right away.

**Comment by zbpetersbuf** [comment]
@ojha-aditya got to resolve conflicts

**Comment by ojha-aditya** [comment]
> @ojha-aditya got to resolve conflicts

Done.

File: calculus.py
Comment (SchrodingersStruggle) [inline calculus.py:281]: I like the error handling on the first step of the function, good way to prevent nesting. Not a request for a change or anything, just a comment.

File: test_calculus.py
Comment (SchrodingersStruggle) [inline test_calculus.py:70]: This unit testing could be more robust. It would be good to test if your function handles the exception cases well, such as a test for if your function handles the cases where the integrand is infinite at any point, wether upper limit, lower limit, or in-between.

Comment (ojha-aditya) [reply test_calculus.py:84]: Incorporated a unit test for error handling too. Compared with the expected output, the test passes. Good suggestion, thanks. I had kept the unit test simple as recalled @laserlab discussing in a class to keep the unit tests minimal.

**Comment by ojha-aditya** [comment]
> All-in-all, good formatting. Exceptionally detailed docstrings and comments make it easier to follow. If you decide to implement more robust unit testing, then that would be great to test exceptions, if not then it should still be good. I will not approve or request any changes yet, until you update what your plan is on the unit testing feedback. If you don't have the time to handle it, I will approve the PR to keep the repository flowing.

I think it is a good idea, I made a function in the unit test module that raises the DivideByZero exception and implemented the unit test for that. It runs successfully, merged that commit.

**Comment by SchrodingersStruggle** [comment]
> > All-in-all, good formatting. Exceptionally detailed docstrings and comments make it easier to follow. If you decide to implement more robust unit testing, then that would be great to test exceptions, if not then it should still be good. I will not approve or request any changes yet, until you update what your plan is on the unit testing feedback. If you don't have the time to handle it, I will approve the PR to keep the repository flowing.
> 
> I think it is a good idea, I made a function in the unit test module that raises the DivideByZero exception and implemented the unit test for that. It runs successfully, merged that commit.

The unit testing now shows the code handles a divide by 0 error on the lower bound well though, nice job

**Review by zbpetersbuf — APPROVED** [review]
looks good

**Comment by zbpetersbuf** [comment]
@SchrodingersStruggle please approve of the pull request if you approve

**Comment by zbpetersbuf** [comment]
@ojha-aditya please remember to update branch

**Comment by ojha-aditya** [comment]
> @ojha-aditya please remember to update branch

Updated now.

**Comment by zbpetersbuf** [comment]
@SchrodingersStruggle please review this this pull request, you have time since this is due on Friday but please as soon as you can

**Comment by SchrodingersStruggle** [comment]
@zbpetersbuf I already did my review for the code. Initially I suggested @ojha-aditya  having a test case beyond just singularity at a lower bound, but since the code still functions without it I'll approve the PR. Also, my concern about adding an increment to the upper bound can be disregarded, I misunderstood how the "bounds" actually act.

**Review by SchrodingersStruggle — APPROVED** [review]
Approving as thorough testing isn't really a necessity.

**Comment by zbpetersbuf** [comment]
@SchrodingersStruggle  It may not be needed, but then I don't have to over ride the push request, if I approved it and accepted the pull request before you approved it you wouldn't have gotten the points

**Comment by zbpetersbuf** [comment]
@SchrodingersStruggle at least that's what I believe @laserlab told me

**Comment by laserlab** [comment]
Yes it makes sense to have a second review always by the maintainer to make sure the PR complies eg file names before merging

**Pull request merged**
